stdlib/Std/src/Foreign/C/Value.luna

```diff
@@ -666,3 +666,58 @@ class Pointer a:
     def free: self.ptr.free
 
 def nullPtr: primNullPtr
+
+native class ForeignPtr a:
+    def toCArg: primForeignPtrToCArg self
+    def cast: primForeignPtrCast self
+    def == that: primForeignPtrEq self that
+    def isNull: self == primNullForeignPtr
+    def moveBytes i: primForeignPtrPlus self i
+    def free: primForeignPtrFree self
+    def _ptr: primForeignPtrToPtr self
+
+class ManagedPointer a:
+    ManagedPointer    a
+    ManagedPointerVal a (ForeignPtr a)
+
+    def typeRep: case self of
+        ManagedPointer a: a
+        ManagedPointerVal a _: a
+
+    def ptr: case self of
+        ManagedPointerVal _ p: p._ptr
+
+    def fptr: case self of
+        ManagedPointerVal _ p: p
+
+    def fromPointer finalizer p:
```

Comment:
Do we want a version of this that provides a default finaliser?

Reply:
I think there is no good choice for a default here. If you want a ManagedPointer to a malloced memory, you can use ManagedPointer.malloc, no need for `fromPointer`.